src/core/utils/validate-ipni-advertisement.ts

```diff
@@ -126,3 +284,91 @@ export async function validateIPNIAdvertisement(
     })
   })
 }
+
+/**
+ * Convert a PDP service URL to an IPNI multiaddr format.
+ *
+ * Storage providers expose their PDP (Proof of Data Possession) service via HTTP/HTTPS
+ * endpoints (e.g., "https://provider.example.com:8443"). When they advertise content
+ * to IPNI, they include multiaddrs in libp2p format (e.g., "/dns/provider.example.com/tcp/8443/https").
+ *
+ * This function converts between these representations to enable validation that a
+ * provider's IPNI advertisement matches their registered service endpoint.
+ *
+ * @param serviceURL - HTTP/HTTPS URL of the provider's PDP service
+ * @param logger - Optional logger for warnings
+ * @returns Multiaddr string in libp2p format, or undefined if conversion fails
+ *
+ * @example
+ * serviceURLToMultiaddr('https://provider.example.com')
+ * // Returns: '/dns/provider.example.com/tcp/443/https'
+ *
+ * @example
+ * serviceURLToMultiaddr('http://provider.example.com:8080')
+ * // Returns: '/dns/provider.example.com/tcp/8080/http'
+ */
+export function serviceURLToMultiaddr(serviceURL: string, logger?: Logger): string | undefined {
+  try {
+    const url = new URL(serviceURL)
+    const port = url.port || (url.protocol === 'https:' ? '443' : '80')
+    const protocolComponent = url.protocol.replace(':', '')
+
+    return `/dns/${url.hostname}/tcp/${port}/${protocolComponent}`
+  } catch (error) {
+    logger?.warn({ serviceURL, error }, 'Unable to derive IPNI multiaddr from serviceURL')
+    return undefined
+  }
+}
+
+/**
+ * Derive expected IPNI multiaddrs from provider information.
+ *
+ * For each provider, attempts to extract their PDP serviceURL and convert it to
+ * the multiaddr format used in IPNI advertisements. This allows validation that
+ * specific providers have advertised the content.
+ *
+ * Note: ProviderInfo should contain the serviceURL at `products.PDP.data.serviceURL`.
+ *
+ * @param providers - Array of provider info objects from synapse SDK
+ * @param extraMultiaddrs - Additional multiaddrs to include in expectations
```

Comment:
What's the usecase of supporting `extraMultiaddrs`?

Reply:
removed. it was for cases where derivedMultiaddrs wasn't right and folks wanted more control. we can add back if needed.